kw/platform/wsgi.py

```diff
@@ -0,0 +1,78 @@
+import re
+import time
+from datetime import datetime
+
+import webob
+from dateutil.parser import parse
+from webob.dec import wsgify
+
+from . import settings
+
+
+user_agent_re = re.compile(settings.VALID_USER_AGENT_RE)
+req_slowdown_datetime = parse(settings.REQUESTS_SLOWDOWN_DATETIME)
+req_restrict_datetime = parse(settings.REQUESTS_RESTRICT_DATETIME)
+
+
+def _refuse_request(req, app):
+    return webob.exc.HTTPBadRequest("Invalid User-Agent")
+
+
+def _slowdown_request(req, app):
+    before_time = time.time()
+    resp = req.get_response(app)
+    seconds = time.time() - before_time
+    time.sleep(seconds)
+    return resp
+
+
+@wsgify.middleware
+def user_agent_middleware(req, app):
+    """Validate client's User-Agent header and modify response based on that.
+
+    If the User-Agent header is invalid, there are three things that can happen:
+
+    1. The current time is less then :obj:`settings.REQUESTS_SLOWDOWN_DATETIME`,
+       do nothing in this case.
+    2. The current time is less then :obj:`settings.REQUESTS_RESTRICT_DATETIME`,
+       slow down the response twice the normal responce time.
+    3. The current time is more then :obj:`settings.REQUESTS_RESTRICT_DATETIME`,
+       refuse the request, return HTTP 400 to the client.
+
+    Usage:
+
+        from your_app import wsgi_app
+
+        wsgi_app = user_agent_middleware(wsgi_app)
+
+    For example, in Flask, the middleware can be applied like this:
+
+        from flask import Flask
+
+        app = Flask(__name__)
+        app.wsgi_app = user_agent_middleware(app)
+
+        app.run()
+
+    In Django, you can apply the middleware like this:
+
+        from django.core.wsgi import get_wsgi_application
+
+        application = user_agent_middleware(get_wsgi_application())
+
+    For more information see
+    `Django docs <https://docs.djangoproject.com/en/dev/howto/deployment/wsgi/>`_.
```

Comment:
Can we add a sphinx-compatible warning block about how slowdowns can increase worker business?